test/harry/main/org/apache/cassandra/harry/model/ASTSingleTableModel.java

```diff
@@ -341,61 +380,142 @@ else if (!clusterings.isEmpty())
-        if (partition == null) return SelectResult.ordered(factory.selectionOrder, NO_ROWS);
-
-        var cd = cdOrNull(mutation);
-        var row = cd == null ? null : partition.get(cd);
-        ImmutableUniqueList<Symbol> columns = cd != null ? factory.selectionOrder : factory.partitionAndStaticColumns;
-        return SelectResult.ordered(columns, new ByteBuffer[][] { getRowAsByteBuffer(columns, partition, row)});
+        BytesPartitionState.Ref ref = referencePartition(mutation);
+        Clustering<ByteBuffer> cd = cdOrNull(mutation);
+        BytesPartitionState partition = partitions.get(ref);
+        return new CasContext(ref, cd, partition);
     }
 
-    private boolean shouldApply(Mutation mutation, SelectResult current)
+    private boolean shouldApply(Mutation mutation, CasContext ctx)
     {
         Preconditions.checkArgument(mutation.isCas());
         // process condition
-        CasCondition condition;
-        switch (mutation.kind)
-        {
-            case INSERT:
-                condition = CasCondition.Simple.NotExists;
-                break;
-            case UPDATE:
-                condition = ((Mutation.Update) mutation).casCondition.get();
-                break;
-            case DELETE:
-                condition = ((Mutation.Delete) mutation).casCondition.get();
-                break;
-            default:
-                throw new UnsupportedOperationException(mutation.kind.name());
-        }
+        CasCondition condition = mutation.casCondition().get();
+        boolean partitionOrRow = ctx.clustering == null;
+        boolean partitionKnown = ctx.partition != null;
+        BytesPartitionState.Row row = partitionKnown && !partitionOrRow
+                                      ? ctx.partition.get(ctx.clustering)
+                                      : null;
         if (condition instanceof CasCondition.Simple)
         {
-            boolean hasPartition = current.rows.length > 0;
-            boolean partitionOrRow = current.columns.equals(factory.partitionAndStaticColumns);
-            boolean hasRow = partitionOrRow ? hasPartition : current.isAllDefined(factory.clusteringColumns);
+            if (partitionOrRow && factory.staticColumns.isEmpty())
+                throw new AssertionError("Attempted to create a EXISTS condition on partition without static columns; " + mutation.toCQL());
+            // CAS's definition of partition EXISTS isn't based off the partition existing, its based off the static row
+            // existing (aka at least 1 static column exists and is not null).
+            boolean hasPartition = partitionKnown && !ctx.partition.staticRow().isEmpty();
+            boolean hasRow = row != null && !row.isEmpty();
             var simple = (CasCondition.Simple) condition;
             switch (simple)
             {
                 case Exists:
-                    return hasRow;
+                    return partitionOrRow ? hasPartition : hasRow;
```

Comment:
turns out that CAS defines partition `EXISTS` as 

1) partition is defined
2) static row is not empty

The previous logic defined it only as #1 so had to make this change to support the empty static row check